packages/react/src/components/Form/PinField.tsx

```diff
@@ -0,0 +1,157 @@
+import {ChangeEvent, KeyboardEvent, useCallback, useEffect, useRef, useState} from 'react';
+import '@compassion-gds/css/src/components/Form/pin.css';
+
+import { FieldSize, FieldState } from './Helpers';
+
+type Values = string[];
+
+interface PinProps {
+  pattern?: RegExp,
+  autoFocus?: boolean;
+  secret?: boolean;
+  state?: FieldState;
+  onComplete?: (value) => void;
+}
+
+interface PinHookProps extends PinProps {
+  values?: Values;
+  onChange?: (values: Values) => void;
+}
+
+export interface PinFieldProps extends PinProps {
+  label?: string;
+  value?: string | number;
+  length?: number;
+  size?: FieldSize;
+  className?: string;
+}
+
+const PinField = ({
+  label,
+  value,
+  length = 4,
+  state,
+  size,
+  secret = false,
+  autoFocus = false,
+  pattern = DEFAULT_PATTERN,
+  className,
+  onComplete,
+}: PinFieldProps) => {
+  const containerClassNames = ['pin-field', size, className].join(' ');
+  const fieldClassNames = ['form-field', state].join(' ');
+  const labelClassNames = ['form-label', state].join(' ');
+
+  const defaultValues = Array.from({ length }).map((_, i) => value?.toString()[i] || '');
+  const [values, setValues] = useState(defaultValues);
+  const fields = usePinHook({
+    values,
+    autoFocus,
+    secret,
+    pattern,
+    state,
+    onChange: v => setValues(v),
+    onComplete,
+  });
+
+  return (
+    <div className="form-field-row">
+      {label && (
+        <label className={labelClassNames}>
+          {label}
+        </label>
+      )}
+      <div className={containerClassNames}>
+        {fields.map((fieldProps, index) => (
+          <div key={index} className={fieldClassNames}>
+            <input className="form-input" {...fieldProps} />
+          </div>
+        ))}
+      </div>
+    </div>
+  );
+};
+
+PinField.displayName = 'PinField';
+
+const usePinHook = ({
+  values,
+  autoFocus,
+  secret,
+  state,
+  pattern,
+  onChange: onChangeProp,
+  onComplete,
+}: PinHookProps = {}) => {
+  const disabled = state === 'disabled';
+  const error = state === 'error';
+
+  const fieldRefs = useRef<HTMLInputElement[]>(Array(values.length).fill(null));
+
+  const setFocus = useCallback((index = 0) => {
+    fieldRefs.current[index]?.focus();
+  }, []);
+
+  useEffect(() => {
+    if (autoFocus) setFocus();
+  }, [autoFocus, setFocus]);
+
+  const setFieldRef = useCallback((index: number) => (ref: HTMLInputElement) => {
+    fieldRefs.current[index] = ref;
+  }, []);
+
+  const onChange = useCallback((index: number) => (event: ChangeEvent<HTMLInputElement>) => {
+    let { value } = event.target;
+
+    value = value.trim().toUpperCase();
+
+    if (!pattern.test(value)) return;
+
+    const nextValues = values.slice();
+    nextValues[index] = value;
+    onChangeProp(nextValues);
```

Comment:
Why do we have to uppercase the input?

I think it's valid to have both lowercase and uppercase letters if the patterns allows it 
Other places that I use PIN allow for both upper and lowercase characters